File: templates/astro-template/src/lib/firebase.ts
import { initializeApp, cert, getApps, type App } from 'firebase-admin/app';
import { getFirestore, type Firestore } from 'firebase-admin/firestore';
import { config } from './config';

let app: App;
let db: Firestore;

function initializeFirebase(): App {
  if (getApps().length > 0) {
    return getApps()[0];
  }

  const { firebase } = config;

  // When using emulator, no credentials needed
  if (firebase.useEmulator) {
    return initializeApp({ projectId: firebase.projectId });
  }

  // Production: use service account credentials
  if (process.env.GOOGLE_APPLICATION_CREDENTIALS) {
    return initializeApp({
      credential: cert(process.env.GOOGLE_APPLICATION_CREDENTIALS),
      projectId: firebase.projectId,
    });
  }

  // Fallback: ADC (Application Default Credentials) - works in GCP environments
  return initializeApp({ projectId: firebase.projectId });
}

export function getDb(): Firestore {
  if (!db) {
    app = initializeFirebase();
    db = getFirestore(app);
  }
  return db;
}

// Helper for common operations
export const firestore = {
  async getDoc<T>(collection: string, id: string): Promise<T | null> {
    const doc = await getDb().collection(collection).doc(id).get();
    return doc.exists ? (doc.data() as T) : null;
  },

  async getDocs<T>(collection: string): Promise<T[]> {
    const snapshot = await getDb().collection(collection).get();
    return snapshot.docs.map((doc) => ({ id: doc.id, ...doc.data() }) as T);
  },

  async addDoc<T extends Record<string, unknown>>(collection: string, data: T): Promise<string> {
    const ref = await getDb().collection(collection).add(data);
    return ref.id;
  },

  async updateDoc<T extends Record<string, unknown>>(collection: string, id: string, data: Partial<T>): Promise<void> {
    await getDb().collection(collection).doc(id).update(data);
  },

  async deleteDoc(collection: string, id: string): Promise<void> {
    await getDb().collection(collection).doc(id).delete();
  },
};

File: templates/flask-template/app/lib/firebase.py
import os
from typing import Any, TypeVar

import firebase_admin
from firebase_admin import credentials, firestore
from google.cloud.firestore_v1 import Client

from app.config import config

_db: Client | None = None

T = TypeVar("T", bound=dict[str, Any])


def _initialize_firebase() -> firebase_admin.App:
    if firebase_admin._apps:
        return firebase_admin.get_app()

    firebase_config = config.firebase

    # When using emulator, no credentials needed
    if firebase_config.use_emulator:
        return firebase_admin.initialize_app(
            options={"projectId": firebase_config.project_id}
        )

    # Production: use service account credentials
    creds_path = os.getenv("GOOGLE_APPLICATION_CREDENTIALS")
    if creds_path:
        cred = credentials.Certificate(creds_path)
        return firebase_admin.initialize_app(
            cred, options={"projectId": firebase_config.project_id}
        )

    # Fallback: ADC (Application Default Credentials) - works in GCP environments
    return firebase_admin.initialize_app(
        options={"projectId": firebase_config.project_id}
    )


def get_db() -> Client:
    global _db
    if _db is None:
        _initialize_firebase()
        _db = firestore.client()
    return _db


class FirestoreHelper:
    """Helper class for common Firestore operations."""

    @staticmethod
    def get_doc(collection: str, doc_id: str) -> dict[str, Any] | None:
        doc = get_db().collection(collection).document(doc_id).get()
        return doc.to_dict() if doc.exists else None

    @staticmethod
    def get_docs(collection: str) -> list[dict[str, Any]]:
        docs = get_db().collection(collection).stream()
        return [{"id": doc.id, **doc.to_dict()} for doc in docs]

    @staticmethod
    def add_doc(collection: str, data: dict[str, Any]) -> str:
        _, ref = get_db().collection(collection).add(data)
        return ref.id

    @staticmethod
    def update_doc(collection: str, doc_id: str, data: dict[str, Any]) -> None:
        get_db().collection(collection).document(doc_id).update(data)

    @staticmethod
    def delete_doc(collection: str, doc_id: str) -> None:
        get_db().collection(collection).document(doc_id).delete()


db = FirestoreHelper()
